src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/AddressBinder.cs

```diff
@@ -0,0 +1,307 @@
+            }
+
+            // If StartLocalhost doesn't throw, there is at least one listener.
+            // The port cannot change for "localhost".
+            context.Addresses.Add(address.ToString());
+        }
+
+        private static async Task BindAddressAsync(string address, AddressBindContext context)
+        {
+            var parsedAddress = ServerAddress.FromUrl(address);
+
+            if (parsedAddress.Scheme.Equals("https", StringComparison.OrdinalIgnoreCase))
+            {
+                throw new InvalidOperationException($"HTTPS endpoints can only be configured using {nameof(KestrelServerOptions)}.{nameof(KestrelServerOptions.Listen)}().");
+            }
+            else if (!parsedAddress.Scheme.Equals("http", StringComparison.OrdinalIgnoreCase))
+            {
+                throw new InvalidOperationException($"Unrecognized scheme in server address '{address}'. Only 'http://' is supported.");
+            }
+
+            if (!string.IsNullOrEmpty(parsedAddress.PathBase))
+            {
+                throw new InvalidOperationException($"A path base can only be configured using {nameof(IApplicationBuilder)}.UsePathBase().");
+            }
+
+            if (parsedAddress.IsUnixPipe)
+            {
+                var endPoint = new ListenOptions(parsedAddress.UnixPipePath);
+                await BindEndpointAsync(endPoint, context);
+                context.Addresses.Add(endPoint.GetDisplayName());
+            }
+            else if (string.Equals(parsedAddress.Host, "localhost", StringComparison.OrdinalIgnoreCase))
+            {
+                // "localhost" for both IPv4 and IPv6 can't be represented as an IPEndPoint.
+                await BindLocalhostAsync(parsedAddress, context);
+            }
+            else
+            {
+                ListenOptions options;
+                if (TryCreateIPEndPoint(parsedAddress, out var endpoint))
+                {
+                    options = new ListenOptions(endpoint);
+                    await BindEndpointAsync(options, context);
+                }
+                else
+                {
+                    // when address is 'http://hostname:port', 'http://*:port', or 'http://+:port'
+                    try
+                    {
+                        options = new ListenOptions(new IPEndPoint(IPAddress.IPv6Any, parsedAddress.Port));
+                        await BindEndpointAsync(options, context);
+                    }
+                    catch (Exception ex) when (!(ex is IOException))
+                    {
+                        // for machines that do not support IPv6
+                        options = new ListenOptions(new IPEndPoint(IPAddress.Any, parsedAddress.Port));
+                        await BindEndpointAsync(options, context);
+                    }
+                }
+
+                context.Addresses.Add(options.GetDisplayName());
+            }
+        }
+
+        private interface IBindStrategy
```

Comment:
The name of your chosen pattern is leaking into your API names, viraly. IBinder, DefaultAddressBinder, etc..?